tfhe/src/integer/server_key/radix_parallel/scalar_bitwise_op.rs

```diff
@@ -0,0 +1,390 @@
+        }
+    }
+
+    pub fn unchecked_scalar_bitor_parallelized<T>(
+        &self,
+        lhs: &RadixCiphertext,
+        rhs: T,
+    ) -> RadixCiphertext
+    where
+        T: DecomposableInto<u8>,
+    {
+        let mut result = lhs.clone();
+        self.unchecked_scalar_bitor_assign_parallelized(&mut result, rhs);
+        result
+    }
+
+    pub fn unchecked_scalar_bitor_assign_parallelized<T>(&self, lhs: &mut RadixCiphertext, rhs: T)
+    where
+        T: DecomposableInto<u8>,
+    {
+        let message_modulus = self.key.message_modulus.0;
+        assert!(message_modulus.is_power_of_two());
+
+        let clear_blocks = BlockDecomposer::with_early_stop_at_zero(rhs, message_modulus.ilog2())
+            .iter_as::<u8>()
+            .collect::<Vec<_>>();
+
+        lhs.blocks
+            .par_iter_mut()
+            .zip(clear_blocks.par_iter().copied())
+            .for_each(|(lhs_block, clear_block)| {
+                self.key
+                    .unchecked_scalar_bitor_assign(lhs_block, clear_block);
+            });
+
+        // Blocks beyond clear_blocks.len() should be 'ored'
+        // with '0', which means they keep their value
+    }
+    pub fn smart_scalar_bitand_parallelized<T>(
+        &self,
+        lhs: &mut RadixCiphertext,
+        rhs: T,
+    ) -> RadixCiphertext
+    where
+        T: DecomposableInto<u8>,
+    {
+        if !lhs.block_carries_are_empty() {
+            self.full_propagate(lhs);
+        }
+        let mut result = lhs.clone();
+        self.unchecked_scalar_bitand_assign_parallelized(&mut result, rhs);
+        result
+    }
+
+    pub fn smart_scalar_bitand_assign_parallelized<T>(&self, lhs: &mut RadixCiphertext, rhs: T)
+    where
+        T: DecomposableInto<u8>,
+    {
+        if !lhs.block_carries_are_empty() {
+            self.full_propagate(lhs);
+        }
+        self.unchecked_scalar_bitand_assign_parallelized(lhs, rhs);
+    }
+
+    /// Computes homomorphically a bitand between a ciphertexts and a clear value
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2;
+    ///
+    /// // Generate the client key and the server key:
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2, num_blocks);
+    ///
+    /// let msg1 = 14u8;
+    /// let msg2 = 97u8;
+    ///
+    /// let ct1 = cks.encrypt(msg1);
+    ///
+    /// let ct_res = sks.scalar_bitand_parallelized(&ct1, msg2);
+    ///
+    /// // Decrypt:
+    /// let dec_result: u8 = cks.decrypt(&ct_res);
+    /// assert_eq!(dec_result, msg1 & msg2);
+    /// ```
+    pub fn scalar_bitand_parallelized<T>(&self, lhs: &RadixCiphertext, rhs: T) -> RadixCiphertext
+    where
+        T: DecomposableInto<u8>,
+    {
+        let mut result = lhs.clone();
+        self.unchecked_scalar_bitand_assign_parallelized(&mut result, rhs);
+        result
```

Comment:
looks to me we don't manage carries properly ? All default ops might have the "defect" in integer/shortint, I let you do a pass and verify we are all good on that front using the corresponding assign implementation and making sure the assign version clears carries.

Maybe some of our default tests need to produce dirty carries (I think some of them do, but I'm not sure) with a check verifying the carries are not empty before the function call and checking they are empty afterwards in addition to returning the right result

Reply:
It should be calling `scalar_bitand_assign_parallelized`